core/src/components/datetime/datetime.tsx

```diff
@@ -2094,16 +2145,21 @@ export class Datetime implements ComponentInterface {
             );
 
             const dateIsoString = convertDataToISO(referenceParts);
-            let isCalDayDisabled = isCalMonthDisabled || disabled;
 
-            if (!isCalDayDisabled && isDateEnabled !== undefined) {
+            /**
+             * Days can be constrained by the min/max bounds or
+             * allowed dates. Constrained days may not be selected.
+             */
+            let isCalDayConstrained = isCalMonthDisabled || isDayDisabled;
```

Comment:
~ Nit: I wouldn't know what constrained means for a variable without looking at this comment. If the behavior is around the day being selected, a name focused around that may be more appropriate.

Reply:
I was going off of "date constraints" which is how the docs refer to this feature. I couldn't think of any better way to distinguish between 

- dates that are specifically disabled (e.g. by min & max dates or specific allowed values) and 
- dates that are disabled for any reason (which includes the above but also if the component is disabled or readonly)

which is what I need to do here.

Reply:
After looking at the code, I think we can do this with a lot less modifications. But please double-check with the behavior you expect:

Revert this back:
```ts
let isCalDayDisabled = isCalMonthDisabled || isDayDisabled;

if (!isCalDayDisabled && isDateEnabled !== undefined) {
  try {
    /**
     * The `isDateEnabled` implementation is try-catch wrapped
     * to prevent exceptions in the user's function from
     * interrupting the calendar rendering.
     */
    isCalDayDisabled = !isDateEnabled(dateIsoString);
  } catch (e) {
    printIonError(
      'Exception thrown from provided `isDateEnabled` function. Please check your function and try again.',
      el,
      e
    );
  }
}
```

Add a boolean for the datetime being "disabled", you call this `isButtonDisabled` in other places, but here I think that could be confusing, so I opted for:
```ts
const isDatetimeDisabled = disabled || readonly;
```

You can then update your JSX bindings (I'll remove all the existing props that are not effected):

```jsx
<button
   disabled={isCalDayDisabled || isDatetimeDisabled}
  class={{
    'calendar-day-constrained': isCalDayDisabled && isDatetimeDisabled,
  }}
```

This should treat the button as disabled either if it is explicitly disabled **or** if the entire datetime is readonly and/or disabled. However it will optically be dimmed in readonly mode only if day is both disabled **and** the datetime is readonly and/or disabled. 

Advantages to this approach is we aren't coupling this constrained state to the disabled state of the day.

Reply:
i like that generally. but some of the specifics are confusing me. 

from other feedback on this PR, i've come to understand that our code style is to have logic in consts, not inside the JSX.

so instead of

```
<button
   disabled={isCalDayDisabled || isDatetimeDisabled}
  class={{
    'calendar-day-constrained': isCalDayDisabled && isDatetimeDisabled,
  }}
```

it should be e.g.

```
<button
   disabled={isButtonDisabled}
  class={{
    'calendar-day-constrained': isFoo,
  }}
```

i updated the PR with both your feedback and the other feedback in mind. still open to ideas for improved const names.

Reply:
There are a few places where abstracting inline expressions into a constant are useful:
1. The expression is re-used in multiple places in the implementation.
2. The expression is needing to access a value off the instance (`this.variableName`). 
3. The expression is either complicated or could benefit from an explicit variable to describe what the check is responsible for. 

I think it is fine either way here if you want to have new constants for abstracting `isCalDayDisabled && isDatetimeDisabled`. I mainly wanted to avoid intermixing the logic for checking if a date is enabled, with the behavior of making the day display at full opacity when in readonly mode.